ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs

```diff
@@ -45,9 +52,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
 import           Ouroboros.Consensus.HardFork.Combinator.Serialisation.Common
 import           Ouroboros.Consensus.HardFork.Combinator.Serialisation.SerialiseDisk ()
+import qualified Ouroboros.Consensus.HardFork.History as History
+
+{-------------------------------------------------------------------------------
+  Serialisation of products
+-------------------------------------------------------------------------------}
+
+-- | Encoding of `NP f xs` while filtering out the components for disabled eras
+-- as indicated by the `HardForkNodeToClientVersion xs`. disabled eras imply
+-- that the protocol version does not support those eras. Hence, omitting the
+-- corresponding elements is the correct behavior.
+encodeNodeToClientNP ::
+     forall xs f. (SerialiseHFC xs)
+  => (   forall x. SerialiseConstraintsHFC x
+      => CodecConfig x
+      -> BlockNodeToClientVersion x
+      -> f x
+      -> Encoding
+     )
+  -- ^ The encoding of the individual elements (assuming era `x` is enabled).
+  -> CodecConfig (HardForkBlock xs)
+  -> HardForkNodeToClientVersion xs
+  -> NP f xs
+  -> Encoding
+encodeNodeToClientNP
+  encodeElement
+  (HardForkCodecConfig (PerEraCodecConfig ccfgs))
+  version
+  xs
+  = case version of
+    HardForkNodeToClientDisabled versionX -> case ccfgs of
+      (ccfg :: CodecConfig x) :* _ -> case xs of
+        x :* _ -> encodeElement ccfg versionX x
+    HardForkNodeToClientEnabled _ subVersions ->
+      let components :: [Encoding]
+          components = catMaybes
+            $ hcollapse
+            $ hczipWith3
+                (Proxy @SerialiseConstraintsHFC)
+                (\ccfg subVersionMay x -> K $ case subVersionMay of
+                  EraNodeToClientEnabled subVersion -> Just (encodeElement ccfg subVersion x)
+                  -- Omit disabled eras
+                  EraNodeToClientDisabled -> Nothing
+                )
+                ccfgs
+                subVersions
+                xs
+          listLen = fromIntegral (length components)
+      in Enc.encodeListLen listLen <> mconcat components
+
+-- | Decoding of `NP f xs`. If any eras are disabled in the version and hence
+-- missing in the serialisation, then this throws an exception. In effect,
+-- deserialisation of product types is only supported when the sender uses an
+-- equal or superset of eras.
+decodeNodeToClientNP ::
+     forall xs f. SerialiseHFC xs
+  => (   forall x. SerialiseConstraintsHFC x
+      => CodecConfig x
+      -> BlockNodeToClientVersion x
+      -> (forall s. Decoder s (f x))
+     )
+  -- ^ The decoding of the individual elements (assuming era `x` is enabled).
+  -> CodecConfig (HardForkBlock xs)
+  -> HardForkNodeToClientVersion xs
+  -> (forall s. Decoder s (NP f xs))
+decodeNodeToClientNP
+  decodeElement
+  (HardForkCodecConfig (PerEraCodecConfig ccfgs))
+  version
+  = case version of
+    HardForkNodeToClientDisabled versionX -> case ccfgs of
```

Comment:
Please indent these alternatives (ie all lines between the `case version of` and `where` lines) two more spaces.